--- Makefile.orig	2008-12-13 14:52:36.000000000 +0800
+++ Makefile	2008-12-13 15:10:17.000000000 +0800
@@ -53,7 +53,7 @@
 # for diffusion
 #DIFF_INCLUDES = "./diffusion3/main ./diffusion3/lib ./diffusion3/nr ./diffusion3/ns"
 
-CCOPT	=  -Wall 
+CCOPT	= -g -Wall -Wno-write-strings
 STATIC	= 
 #LDFLAGS	= $(STATIC)
 LDFLAGS	=  -Wl,-export-dynamic 
@@ -323,6 +323,7 @@
 	wpan/p802_15_4sscs.o wpan/p802_15_4timer.o \
 	wpan/p802_15_4trace.o wpan/p802_15_4transac.o \
 	apps/pbc.o \
+	src/mflood/mflood.o src/mflood/mflood-seqtable.o \
 	$(OBJ_STL)
 
 
--- tcl/lib/ns-lib.tcl.orig	2008-12-13 14:45:42.000000000 +0800
+++ tcl/lib/ns-lib.tcl	2008-12-14 14:29:11.000000000 +0800
@@ -621,6 +621,9 @@
 		set ragent [$self $rtAgentFunction_ $node]
 	} else {
 		switch -exact $routingAgent_ {
+		    MFlood {
+			    set ragent [$self create-mflood-agent $node]
+		    }
 		    DSDV {
 			    set ragent [$self create-dsdv-agent $node]
 		    }
@@ -791,6 +794,13 @@
 	return $nodetype
 }
 
+#mflood
+Simulator instproc create-mflood-agent { node } {
+	set ragent [new Agent/MFlood [$node id]]
+	$node set ragent_ $ragent
+	return $ragent
+}
+
 Simulator instproc create-tora-agent { node } {
         set ragent [new Agent/TORA [$node id]]
 	$node set ragent_ $ragent
--- tcl/lib/ns-packet.tcl.orig	2008-12-14 11:26:18.000000000 +0800
+++ tcl/lib/ns-packet.tcl	2008-12-14 11:27:17.000000000 +0800
@@ -116,6 +116,8 @@
 	Flags
 	IP 	# IP
 # Routing Protocols:
+	#mflood
+	MFlood
 	NV 	# NixVector classifier for stateless routing 
 	rtProtoDV 	# distance vector routing protocol
 	rtProtoLS 	# link state routing protocol
--- common/packet.h.orig	2008-12-13 15:20:23.000000000 +0800
+++ common/packet.h	2008-12-13 16:03:18.000000000 +0800
@@ -82,7 +82,6 @@
  */
 
 typedef unsigned int packet_t;
-
 static const packet_t PT_TCP = 0;
 static const packet_t PT_UDP = 1;
 static const packet_t PT_CBR = 2;
@@ -180,7 +179,8 @@
 static const packet_t PT_BLTRACE = 60;
 
         // insert new packet types here
-static packet_t       PT_NTYPE = 61; // This MUST be the LAST one
+static const packet_t PT_MFLOOD = 61;
+static packet_t       PT_NTYPE = 62; // This MUST be the LAST one
 
 enum packetClass
 {
@@ -376,6 +376,9 @@
 
 		// Bell Labs (PackMime OL)
 		name_[PT_BLTRACE]="BellLabsTrace";
+
+		//mflood
+		name_[PT_MFLOOD] = "MFlood";
 		
 		name_[PT_NTYPE]= "undefined";
 	}
